src/array_api_stubs/_draft/creation_functions.py

```diff
@@ -214,19 +214,36 @@ def eye(
     """
 
 
-def from_dlpack(x: object, /) -> array:
+def from_dlpack(
+    x: object,
+    /,
+    *,
+    device: Optional[device] = None,
+    copy: Optional[bool] = None,
+) -> array:
     """
     Returns a new array containing the data from another (array) object with a ``__dlpack__`` method.
 
     Parameters
     ----------
     x: object
         input (array) object.
+    device: Optional[device]
+        device on which to place the created array. If ``device`` is ``None`` and ``x`` supports DLPack, the output array must be on the same device as ``x``. Default: ``None``.
+
+        The v2023.12 standard only mandates that a compliant library should offer a way for ``from_dlpack`` to create an array
+        whose underlying memory is accessible to the Python interpreter (represented by the ``kDLCPU`` enumerator in DLPack).
```

Comment:
I think this is still in the wrong place. The requirement should be for `__dlpack__` to offer a way to export a DLPack capsule with data on host memory. `from_dlpack` doesn't have to allow this (we cannot mandate CuPy to return a CPU array to the user).

The sentence is fine as is, as long as `from_dlpack` is replaced by `__dlpack__` and moved to the `__dlpack__` docstring.

Reply:
I think the array API users should know about this limitation too, no?

Reply:
Ah, I see what you mean. It makes sense to add something here too if you're intending to leave the option open to not support non-CPU devices as inputs. Which is one of two ways to go here.

So then:
- `__dlpack__` should say "must support kDLCPU"
- `from_dlpack` should say "must support returning an array whose underlying memory is accessible to the Python interpreter _if the library has such an array type_" and "may support non-host-memory devices in this API (if it does, data transfer may go via host memory)".

It does make it more complicated to explain, but if you don't want to make everything unconditionally work in `from_dlpack` to avoid mandating double copies in some cases, that's the price to pay.

Reply:
> I think the array API users should know about this limitation too, no?

True, but maybe we need to be more specific maybe:
> As of the v2023.12 standard, a successful copy to `dl_device` is only guaranteed to be implemented when the target `dl_device` represents host (CPU) memory accessible to the Python interpreter (i.e. `np.from_dlpack(x, device="cpu")` must work; however, this doesn't apply `cp.from_dlpack(x, device="cpu")` as cupy does not have a "cpu" device)

I dunno about the host (CPU), the "accessible from the Python interpreter" is unfortunately a pretty abstract mouthful :).

Reply:
Nvm. Ralfs version is shorter and better :).  I might change "must" -> "is guaranteed to".

Reply:
Let me know if commit 24cb4db works for you. I don't wanna use the wording "if the library has such an array type" because it's a bit confusing. Also, I don't mention "non-host-memory" to keep things simple.

Reply:
That looks reasonable. One question:

> If the array library does not support ``kDLCPU`` at all,

Should that phrasing be removed from the `__dlpack__` description? Is there a realistic chance that there will be a library that really cannot copy data to host memory? If not, I don't think it is necessary to hedge there.

Reply:
It was my attempt to mirror the `device` treatment in `from_dlpack`, but I see your point. A library does not necessarily have a `device` object representing CPU, but supporting it at the C level is no problem. Removed in commit 43263ce.